prov/efa/src/efa_cq.h

```diff
@@ -176,3 +184,107 @@ static inline int efa_cq_ibv_cq_ex_open(struct fi_cq_attr *attr,
 		&init_attr_ex, ibv_ctx, ibv_cq_ex, ibv_cq_ex_type);
 }
 #endif
+
+int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
+		struct fid_cq **cq_fid, void *context);
+
+void efa_cq_progress(struct util_cq *cq);
+
+#if HAVE_CAPS_UNSOLICITED_WRITE_RECV
+/**
+ * @brief Check whether a completion consumes recv buffer
+ *
+ * @param ibv_cq_ex extended ibv cq
+ * @return true the wc consumes a recv buffer
+ * @return false the wc doesn't consume a recv buffer
+ */
+static inline
+bool efa_cq_wc_is_unsolicited(struct ibv_cq_ex *ibv_cq_ex)
+{
+	return efa_use_unsolicited_write_recv() && efadv_wc_is_unsolicited(efadv_cq_from_ibv_cq_ex(ibv_cq_ex));
+}
+
+#else
+
+static inline
+bool efa_cq_wc_is_unsolicited(struct ibv_cq_ex *ibv_cq_ex)
+{
+	return false;
+}
+
+#endif
+
+static inline const char *efa_ep_raw_addr_str(struct efa_base_ep *base_ep, char *buf, size_t *buflen)
+{
+	return ofi_straddr(buf, buflen, FI_ADDR_EFA, &base_ep->src_addr);
+}
+
+static inline const char *efa_ep_get_peer_raw_addr_str(struct efa_base_ep *base_ep, fi_addr_t addr, char *buf, size_t *buflen)
+{
+	struct efa_conn *efa_conn;
+	efa_conn = efa_av_addr_to_conn(base_ep->av, addr);
+
+	return ofi_straddr(buf, buflen, FI_ADDR_EFA, efa_conn ? efa_conn->ep_addr : NULL);
+}
+
+/**
+ * @brief Write the error message and return its byte length
+ * @param[in]    ep          EFA base endpoint
+ * @param[in]    addr        Remote peer fi_addr_t
+ * @param[in]    prov_errno  EFA provider * error code(must be positive)
+ * @param[out]   buf         Pointer to the address of error data written by
+ * this function
+ * @param[out]   buflen      Pointer to the returned error data size
+ * @return       A status code. 0 if the error data was written successfully,
+ * otherwise a negative FI error code.
+ */
+static inline int efa_write_error_msg(struct efa_base_ep *ep, fi_addr_t addr,
```

Comment:
I am trying to refactor this function path and make the error msg construction a separate function https://github.com/ofiwg/libfabric/pull/10657. Can you rebase your change and reuse that?

Reply:
Or you can make it general for efa-rdm and efa-raw so I will rebase on your change

Reply:
I am not sure how to do this cleanly since I can't get the peer from base_ep

Reply:
I see, no bother then